aws_analysis_tools/cli/instances_2.py

```diff
@@ -0,0 +1,217 @@
+    _CLI_TO_AWS = {
+            'group': 'group-name',
+            'name': 'tag:Name',
+            'type': 'instance-type',
+            'zone': 'availability-zone',
+            'state': 'instance-state-name'
+        }
+
+    # List of all options
+    _OPTS = ['group', 'name', 'type', 'zone', 'state']
+
+    # A dict with key=CLI options and value=instance attribute
+    _INSTANCE_ATTR = {
+            'group': lambda i: i.groups[0].name,
+            'name': lambda i: i.tags.get('Name'),
+            'type': lambda i: i.instance_type,
+            'zone': lambda i: str(i._placement),
+            'state': lambda i: i.state
+        }
+
+    def __init__(self, name=NAME):
+        # Call to the superclass to bootstrap.
+        super(Application, self).__init__(name=name)
+
+        self.options = vars(self.args)
+
+    # Convert options dictionary to use AWS filters as keys instead of
+    # CLI options
+    def convert_args(self):
+        # Dictionary of options and values to put in the Filter
+        filter_dict = {}
+
+        # Add entries to filter_dict with key=AWS filters and value=option
+        # values for options that filter on inclusion
+        for opt in Application._OPTS:
+            if self.options[opt]:
+                aws_filter = Application._CLI_TO_AWS[opt]
+                filter_dict[aws_filter] = self.options[opt]
+
+        return filter_dict
+
+    # Use filter_dict to filter instances based on inclusion/exclusion options
+    def filter_args(self, filter_dict):
+
+        f = Filter(filter_dict)
+
+        # Filter/find instances based on inclusion filters
+        instances = self.ec2.find_instances(f)
+
+        # Iterate through found instances and filter based on exclude options
+        for opt in Application._OPTS:
+            exclude_str = 'exclude_' + opt
+
+            if self.options[exclude_str]:
+                attribute = self.options[exclude_str]
+            else:
+                continue
+
+            # Exclude instances if they have an attribute that is excluded
+            instances = [
+                            i for i in instances if
+                            Application._INSTANCE_ATTR[opt](i) != attribute
+                        ]
+
+        return instances
+
+    # Outputs filtered instances as a table
+    def output_table(self, instances):
+        table       = Texttable( max_width=0 )
+
+        table.set_deco( Texttable.HEADER )
+        table.set_cols_dtype( [ 't', 't', 't', 't', 't', 't', 't', 't' ] )
+        table.set_cols_align( [ 'l', 'l', 'l', 'l', 'l', 'l', 'l', 't' ] )
+
+        if not self.args.no_header:
+            ### using add_row, so the headers aren't being centered, for easier grepping
+            table.add_row(
+                [ '# id', 'Name', 'Type', 'Zone', 'Group', 'State', 'Root', 'Volumes' ] )
+
+        for i in instances:
+
+            ### XXX there's a bug where you can't get the size of the volumes, it's
```

Comment:
Nit: `###` commenting style doesn't match the rest of the code. Change this to `#` commenting style.